File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/editor/composition/CompositionEditorPresenter.java
package com.github.anrimian.musicplayer.ui.editor.composition;

import com.github.anrimian.musicplayer.domain.interactors.editor.EditorInteractor;
import com.github.anrimian.musicplayer.domain.models.composition.FullComposition;
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre;
import com.github.anrimian.musicplayer.domain.models.image.ImageSource;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import java.util.List;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import moxy.MvpPresenter;

import static com.github.anrimian.musicplayer.data.utils.rx.RxUtils.dispose;


public class CompositionEditorPresenter extends MvpPresenter<CompositionEditorView> {

    private final long compositionId;
    private final EditorInteractor editorInteractor;
    private final Scheduler uiScheduler;
    private final ErrorParser errorParser;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();
    private Disposable changeDisposable;

    private FullComposition composition;

    private ShortGenre removedGenre;

    public CompositionEditorPresenter(long compositionId,
                                      EditorInteractor editorInteractor,
                                      Scheduler uiScheduler,
                                      ErrorParser errorParser) {
        this.compositionId = compositionId;
        this.editorInteractor = editorInteractor;
        this.uiScheduler = uiScheduler;
        this.errorParser = errorParser;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        loadComposition();
        loadGenres();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onChangeAuthorClicked() {
        if (composition == null) {
            return;
        }
        editorInteractor.getAuthorNames()
                .observeOn(uiScheduler)
                .doOnSuccess(artists -> getViewState().showEnterAuthorDialog(composition, artists))
                .doOnError(throwable -> {
                    getViewState().showEnterAuthorDialog(composition, null);
                    onDefaultError(throwable);
                })
                .ignoreElement()
                .onErrorComplete()
                .subscribe();
    }

    void onChangeTitleClicked() {
        if (composition == null) {
            return;
        }
        getViewState().showEnterTitleDialog(composition);
    }

    void onChangeLyricsClicked() {
        if (composition == null) {
            return;
        }
        getViewState().showEnterLyricsDialog(composition);
    }

    void onChangeFileNameClicked() {
        if (composition == null) {
            return;
        }
        getViewState().showEnterFileNameDialog(composition);
    }

    void onChangeAlbumClicked() {
        if (composition == null) {
            return;
        }
        editorInteractor.getAlbumNames()
                .observeOn(uiScheduler)
                .doOnSuccess(albums -> getViewState().showEnterAlbumDialog(composition, albums))
                .doOnError(throwable -> {
                    getViewState().showEnterAlbumDialog(composition, null);
                    onDefaultError(throwable);
                })
                .ignoreElement()
                .onErrorComplete()
                .subscribe();
    }

    void onChangeAlbumArtistClicked() {
        if (composition == null) {
            return;
        }
        editorInteractor.getAuthorNames()
                .observeOn(uiScheduler)
                .doOnSuccess(albums -> getViewState().showEnterAlbumArtistDialog(composition, albums))
                .doOnError(throwable -> {
                    getViewState().showEnterAlbumArtistDialog(composition, null);
                    onDefaultError(throwable);
                })
                .ignoreElement()
                .onErrorComplete()
                .subscribe();
    }

    void onAddGenreItemClicked() {
        editorInteractor.getGenreNames()
                .observeOn(uiScheduler)
                .doOnSuccess(genres -> getViewState().showAddGenreDialog(genres))
                .doOnError(throwable -> {
                    getViewState().showAddGenreDialog(null);
                    onDefaultError(throwable);
                })
                .ignoreElement()
                .onErrorComplete()
                .subscribe();
    }

    void onGenreItemClicked(ShortGenre genre) {
        editorInteractor.getGenreNames()
                .observeOn(uiScheduler)
                .doOnSuccess(genres -> getViewState().showEditGenreDialog(genre, genres))
                .doOnError(throwable -> {
                    getViewState().showEditGenreDialog(genre, null);
                    onDefaultError(throwable);
                })
                .ignoreElement()
                .onErrorComplete()
                .subscribe();
    }

    void onNewGenreEntered(String genre) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.addCompositionGenre(composition, genre)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewGenreNameEntered(String newName, ShortGenre oldGenre) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.changeCompositionGenre(composition, oldGenre, newName)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onRemoveGenreClicked(ShortGenre genre) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.removeCompositionGenre(composition, genre)
                .observeOn(uiScheduler)
                .subscribe(() -> onGenreRemoved(genre), this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onRestoreRemovedGenreClicked() {
        if (removedGenre == null) {
            return;
        }
        onNewGenreEntered(removedGenre.getName());
    }

    void onNewAuthorEntered(String author) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionAuthor(composition, author)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewAlbumEntered(String album) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionAlbum(composition, album)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewAlbumArtistEntered(String artist) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionAlbumArtist(composition, artist)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewTitleEntered(String title) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionTitle(composition, title)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewFileNameEntered(String fileName) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionFileName(composition, fileName)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewLyricsEntered(String text) {
        if (composition == null) {
            return;
        }

        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.editCompositionLyrics(composition, text)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onCopyFileNameClicked() {
        if (composition == null) {
            return;
        }
        getViewState().copyFileNameText(composition.getFileName());
    }

    void onChangeCoverClicked() {
        if (composition == null) {
            return;
        }
        getViewState().showCoverActionsDialog();
    }

    void onClearCoverClicked() {
        if (composition == null) {
            return;
        }
        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.removeCompositionAlbumArt(composition)
                .observeOn(uiScheduler)
                .doOnSubscribe(d -> getViewState().showChangeCoverProgress())
                .doFinally(() -> getViewState().hideChangeCoverProgress())
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    void onNewCoverSelected() {
        getViewState().showSelectImageFromGalleryScreen();
    }

    void onNewImageForCoverSelected(ImageSource imageSource) {
        if (composition == null) {
            return;
        }
        dispose(changeDisposable, presenterDisposable);
        changeDisposable = editorInteractor.changeCompositionAlbumArt(composition, imageSource)
                .observeOn(uiScheduler)
                .doOnSubscribe(d -> getViewState().showChangeCoverProgress())
                .doFinally(() -> getViewState().hideChangeCoverProgress())
                .subscribe(() -> {}, this::onDefaultError);
        presenterDisposable.add(changeDisposable);
    }

    private void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showErrorMessage(errorCommand);
    }

    private void loadComposition() {
        presenterDisposable.add(editorInteractor.getCompositionObservable(compositionId)
                .observeOn(uiScheduler)
                .subscribe(this::onCompositionReceived,
                        this::onCompositionLoadingError,
                        getViewState()::closeScreen));
    }

    private void loadGenres() {
        presenterDisposable.add(editorInteractor.getShortGenresInComposition(compositionId)
                .observeOn(uiScheduler)
                .subscribe(this::onGenresReceived, this::onDefaultError));
    }

    private void onGenresReceived(List<ShortGenre> shortGenres) {
        getViewState().showGenres(shortGenres);
    }

    private void onCompositionReceived(FullComposition composition) {
        if (this.composition == null) {
            checkCompositionTagsInSource(composition);
        }
        this.composition = composition;
        getViewState().showComposition(composition);
    }

    private void checkCompositionTagsInSource(FullComposition composition) {
        presenterDisposable.add(editorInteractor.updateTagsFromSource(composition)
                .observeOn(uiScheduler)
                .subscribe(() -> {}, this::onDefaultError));
    }

    private void onCompositionLoadingError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showCompositionLoadingError(errorCommand);
    }

    private void onGenreRemoved(ShortGenre genre) {
        removedGenre = genre;
        getViewState().showRemovedGenreMessage(genre);
    }
}

File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/editor/composition/CompositionEditorPresenter.kt
package com.github.anrimian.musicplayer.ui.editor.composition

import com.github.anrimian.musicplayer.data.utils.rx.RxUtils
import com.github.anrimian.musicplayer.domain.interactors.editor.EditorInteractor
import com.github.anrimian.musicplayer.domain.models.composition.FullComposition
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre
import com.github.anrimian.musicplayer.domain.models.image.ImageSource
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Scheduler
import io.reactivex.rxjava3.disposables.Disposable

class CompositionEditorPresenter(
        private val compositionId: Long,
        private val editorInteractor: EditorInteractor,
        uiScheduler: Scheduler,
        errorParser: ErrorParser
) : AppPresenter<CompositionEditorView>(uiScheduler, errorParser) {
    
    private var changeDisposable: Disposable? = null
    
    private lateinit var composition: FullComposition
    private var removedGenre: ShortGenre? = null
    private var lastEditAction: Completable? = null
    
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        loadComposition()
        loadGenres()
    }

    fun onChangeAuthorClicked() {
        if (!::composition.isInitialized) {
            return
        }
        editorInteractor.authorNames
                .observeOn(uiScheduler)
                .doOnSuccess { artists -> viewState.showEnterAuthorDialog(composition, artists) }
                .doOnError { throwable ->
                    viewState.showEnterAuthorDialog(composition, null)
                    onDefaultError(throwable)
                }
                .ignoreElement()
                .onErrorComplete()
                .subscribe()
    }

    fun onChangeTitleClicked() {
        if (!::composition.isInitialized) {
            return
        }
        viewState.showEnterTitleDialog(composition)
    }

    fun onChangeLyricsClicked() {
        if (!::composition.isInitialized) {
            return
        }
        viewState.showEnterLyricsDialog(composition)
    }

    fun onChangeFileNameClicked() {
        if (!::composition.isInitialized) {
            return
        }
        viewState.showEnterFileNameDialog(composition)
    }

    fun onChangeAlbumClicked() {
        if (!::composition.isInitialized) {
            return
        }
        editorInteractor.albumNames
                .observeOn(uiScheduler)
                .doOnSuccess { albums -> viewState.showEnterAlbumDialog(composition, albums) }
                .doOnError { throwable ->
                    viewState.showEnterAlbumDialog(composition, null)
                    onDefaultError(throwable)
                }
                .ignoreElement()
                .onErrorComplete()
                .subscribe()
    }

    fun onChangeAlbumArtistClicked() {
        if (!::composition.isInitialized) {
            return
        }
        editorInteractor.authorNames
                .observeOn(uiScheduler)
                .doOnSuccess { albums -> viewState.showEnterAlbumArtistDialog(composition, albums) }
                .doOnError { throwable ->
                    viewState.showEnterAlbumArtistDialog(composition, null)
                    onDefaultError(throwable)
                }
                .ignoreElement()
                .onErrorComplete()
                .subscribe()
    }

    fun onAddGenreItemClicked() {
        editorInteractor.genreNames
                .observeOn(uiScheduler)
                .doOnSuccess(viewState::showAddGenreDialog)
                .doOnError { throwable ->
                    viewState.showAddGenreDialog(null)
                    onDefaultError(throwable)
                }
                .ignoreElement()
                .onErrorComplete()
                .subscribe()
    }

    fun onGenreItemClicked(genre: ShortGenre) {
        editorInteractor.genreNames
                .observeOn(uiScheduler)
                .doOnSuccess { genres -> viewState.showEditGenreDialog(genre, genres) }
                .doOnError { throwable ->
                    viewState.showEditGenreDialog(genre, null)
                    onDefaultError(throwable)
                }
                .ignoreElement()
                .onErrorComplete()
                .subscribe()
    }

    fun onNewGenreEntered(genre: String) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.addCompositionGenre(composition, genre)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewGenreNameEntered(newName: String?, oldGenre: ShortGenre?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.changeCompositionGenre(composition, oldGenre, newName)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onRemoveGenreClicked(genre: ShortGenre) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.removeCompositionGenre(composition, genre)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({ onGenreRemoved(genre) }, this::onDefaultError, presenterDisposable)
    }

    fun onRestoreRemovedGenreClicked() {
        if (removedGenre == null) {
            return
        }
        onNewGenreEntered(removedGenre!!.name)
    }

    fun onNewAuthorEntered(author: String?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionAuthor(composition, author)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewAlbumEntered(album: String?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionAlbum(composition, album)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewAlbumArtistEntered(artist: String?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionAlbumArtist(composition, artist)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewTitleEntered(title: String) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionTitle(composition, title)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewFileNameEntered(fileName: String) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionFileName(composition, fileName)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewLyricsEntered(text: String?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.editCompositionLyrics(composition, text)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onCopyFileNameClicked() {
        if (!::composition.isInitialized) {
            return
        }
        viewState.copyFileNameText(composition.fileName)
    }

    fun onChangeCoverClicked() {
        if (!::composition.isInitialized) {
            return
        }
        viewState.showCoverActionsDialog()
    }

    fun onClearCoverClicked() {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.removeCompositionAlbumArt(composition)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onNewCoverSelected() {
        viewState.showSelectImageFromGalleryScreen()
    }

    fun onNewImageForCoverSelected(imageSource: ImageSource?) {
        if (!::composition.isInitialized) {
            return
        }
        RxUtils.dispose(changeDisposable, presenterDisposable)
        lastEditAction = editorInteractor.changeCompositionAlbumArt(composition, imageSource)
                .observeOn(uiScheduler)
                .doOnSubscribe { viewState.showChangeFileProgress() }
                .doFinally { viewState.hideChangeFileProgress() }
        changeDisposable = lastEditAction!!.subscribe({}, this::onDefaultError, presenterDisposable)
    }

    fun onRetryFailedEditActionClicked() {
        if (lastEditAction != null) {
            RxUtils.dispose(changeDisposable, presenterDisposable)
            changeDisposable = lastEditAction!!
                    .doFinally { lastEditAction = null }
                    .subscribe({}, this::onDefaultError, presenterDisposable)
        }
    }

    private fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showErrorMessage(errorCommand)
    }

    private fun loadComposition() {
        editorInteractor.getCompositionObservable(compositionId)
                .subscribeOnUi(
                        this::onCompositionReceived,
                        this::onCompositionLoadingError,
                        viewState::closeScreen
                )
    }

    private fun loadGenres() {
        editorInteractor.getShortGenresInComposition(compositionId)
                .subscribeOnUi(this::onGenresReceived, this::onDefaultError)
    }

    private fun onGenresReceived(shortGenres: List<ShortGenre>) {
        viewState.showGenres(shortGenres)
    }

    private fun onCompositionReceived(composition: FullComposition) {
        val firstReceive = !this::composition.isInitialized
        if (firstReceive) {
            checkCompositionTagsInSource(composition)
        }
        if (firstReceive || this.composition.dateModified != composition.dateModified) {
            viewState.showCompositionCover(composition)
        }
        this.composition = composition
        viewState.showComposition(composition)
    }

    private fun checkCompositionTagsInSource(composition: FullComposition) {
        editorInteractor.updateTagsFromSource(composition)
                .justSubscribeOnUi(this::onTagCheckError)
    }

    private fun onTagCheckError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showCheckTagsErrorMessage(errorCommand)
    }

    private fun onCompositionLoadingError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showCompositionLoadingError(errorCommand)
    }

    private fun onGenreRemoved(genre: ShortGenre) {
        removedGenre = genre
        viewState.showRemovedGenreMessage(genre)
    }
}